stan/math/prim/prob/generalized_normal_lpdf.hpp

```diff
@@ -0,0 +1,156 @@
+          typename T_shape,
+          require_all_not_nonscalar_prim_or_rev_kernel_expression_t<
+              T_y, T_loc, T_scale, T_shape>* = nullptr>
+inline return_type_t<T_y, T_loc, T_scale, T_shape> generalized_normal_lpdf(
+    T_y&& y, T_loc&& mu, T_scale&& alpha, T_shape&& beta) {
+  using T_partials_return = partials_return_t<T_y, T_loc, T_scale, T_shape>;
+  using T_y_ref = ref_type_if_not_constant_t<T_y>;
+  using T_mu_ref = ref_type_if_not_constant_t<T_loc>;
+  using T_alpha_ref = ref_type_if_not_constant_t<T_scale>;
+  using T_beta_ref = ref_type_if_not_constant_t<T_shape>;
+  static constexpr const char* function = "generalized_normal_lpdf";
+  check_consistent_sizes(function, "Random variable", y, "Location parameter",
+                         mu, "Scale parameter", alpha, "Shape parameter", beta);
+
+  T_y_ref y_ref = std::forward<T_y>(y);
+  T_mu_ref mu_ref = std::forward<T_loc>(mu);
+  T_alpha_ref alpha_ref = std::forward<T_scale>(alpha);
+  T_beta_ref beta_ref = std::forward<T_shape>(beta);
+
+  decltype(auto) y_val = to_ref(as_value_column_array_or_scalar(y_ref));
+  decltype(auto) mu_val = to_ref(as_value_column_array_or_scalar(mu_ref));
+  decltype(auto) alpha_val = to_ref(as_value_column_array_or_scalar(alpha_ref));
+  decltype(auto) beta_val = to_ref(as_value_column_array_or_scalar(beta_ref));
+
+  check_not_nan(function, "Random variable", y_val);
+  check_finite(function, "Location parameter", mu_val);
+  check_positive(function, "Scale parameter", alpha_val);
+  check_positive(function, "Shape parameter",
+                 beta_val);  // With β = +∞ this could be defined to be uniform,
+                             // but we don't support that.
+
+  if (size_zero(y, mu, alpha, beta)) {
+    return 0;
+  }
+  if (!include_summand<propto, T_y, T_loc, T_scale, T_shape>::value) {
+    return 0;
+  }
+
+  const auto& inv_beta1p
+      = to_ref_if<!is_constant<T_shape>::value>(inv(beta_val) + 1);
+  const auto& diff
+      = to_ref_if<!is_constant_all<T_y, T_loc>::value>(y_val - mu_val);
+  const auto& inv_alpha = to_ref(inv(alpha_val));
+  const auto& scaled_abs_diff
+      = to_ref_if<!is_constant_all<T_y, T_loc, T_shape>::value>(abs(diff)
+                                                                * inv_alpha);
+  const auto& scaled_abs_diff_pow
+      = to_ref_if<!is_constant_all<T_scale, T_shape>::value>(
+          pow(scaled_abs_diff, beta_val));
+  const size_t N = max_size(y, mu, alpha, beta);
+
+  T_partials_return logp = -sum(scaled_abs_diff_pow);
+
+  if (include_summand<propto>::value) {
+    logp -= LOG_TWO * N;
+  }
+  if (include_summand<propto, T_scale>::value) {
+    logp -= sum(log(alpha_val)) * (N / math::size(alpha));
+  }
+  if (include_summand<propto, T_shape>::value) {
+    logp -= sum(lgamma(inv_beta1p)) * (N / math::size(beta));
+  }
+
+  auto ops_partials
+      = make_partials_propagator(y_ref, mu_ref, alpha_ref, beta_ref);
+
+  if (!is_constant_all<T_y, T_loc>::value) {
```

Comment:
For any if statements with known compile time values we can use `if constexpr`

```suggestion
  if constexpr (!is_constant_all<T_y, T_loc>::value) {
```